kratos/containers/multi_level_data_value_container.h

```diff
@@ -0,0 +1,472 @@
+                    nullptr;
+                mSize = rOther.mSize;
+
+                const std::size_t allocation_size = mpVariable->Size() / sizeof(BlockType);
+                delete[] mpValues;
+                mpValues = new BlockType[mSize * allocation_size];
+                for (std::size_t i = 0; i < mSize; i++) {
+                    mpVariable->Copy(mpValues + i * allocation_size, rOther.mpValues + i * allocation_size);
+                }
+            }
+            return *this;
+        }
+
+        DataBlock& operator=(DataBlock&& rOther) noexcept
+        {
+            if (this != &rOther) {
+                mpVariable = std::exchange(rOther.mpVariable, nullptr);
+                mpIndexAccesor = std::exchange(rOther.mpIndexAccesor, nullptr);
+                mSize = std::exchange(rOther.mSize, 0);
+                delete[] mpValues;
+                mpValues = std::exchange(rOther.mpValues, nullptr);
+            }
+            return *this;
+        }
+
+        std::size_t size() const
+        {
+            return mSize;
+        }
+
+        const VariableData& GetVariable() const
+        {
+            return *mpVariable;
+        }
+
+        template<typename... Args>
+        std::size_t GetIndex(
+            Args&&... Indices) const
+        {
+            return mpIndexAccesor->GetIndex(std::forward<Args>(Indices)...);
+        }
+
+        template<class TDataType>
+        TDataType* pValue(std::size_t Index)
+        {
+            std::size_t offset = Index * mpVariable->Size()/sizeof(BlockType);
+            TDataType* p = reinterpret_cast<TDataType*>(mpValues + offset);
+            return p;
+        }
+
+        template<class TDataType>
+        const TDataType* pValue(std::size_t Index) const
+        {
+            std::size_t offset = Index * mpVariable->Size()/sizeof(BlockType);
+            const TDataType* p = reinterpret_cast<const TDataType*>(mpValues + offset);
+            return p;
+        }
+
+        const BlockType* pPointer(std::size_t Index) const
+        {
+            std::size_t offset = Index * mpVariable->Size()/sizeof(BlockType);
+            return mpValues + offset;
+        }
+
+        virtual ~DataBlock()
+        {
+            if (mpValues == nullptr) {
+                return;
+            }
+            const std::size_t allocation_size = mpVariable->Size()/sizeof(BlockType);
+            for(std::size_t i = 0; i < mSize; i++) {
+                mpVariable->Destruct(mpValues + i * allocation_size);
+            }
+
+            delete[] mpValues;
+        }
+
+        const MultiLevelDataValueAccesor::UniquePointer& GetIndexAccesor() const
+        {
+            return mpIndexAccesor;
+        }
+
+    private:
+
+        BlockType* Allocate(const VariableData* pVariable , std::size_t Size)
+        {
+            std::size_t allocation_size = pVariable->Size()/sizeof(BlockType);
+            return new BlockType[Size*allocation_size];
+        }
```

Comment:
yikes. How about an `std::unique_ptr`?

btw this seems to be used as a `static` member so why not mark it as such?